[squeezebox] Add LMS favorites channels to server and player

mhilbush opened the pull request:
Signed-off-by: Mark Hilbush <mark@hilbush.com>
Also-by: Gregory Moyer <moyerg@syphr.com>

Add two new channels to enable use of the favorites list on the LMS (Logitech Media Server).

Add a readonly `favoritesList` channel of type String to the squeezebox server thing. 
- the SqueezeBox Server `favoritesList` channel is updated immediately when the binding starts, and whenever the favorites list is updated on the LMS (there’s an asynchronous notification from the LMS to the binding when the favorites list is updated).
- the channel is formatted with comma-separated values of the form favoriteId=favoriteName, like this:
```
1=Rock,2=Folk,3=Classic Alternative
```
- there’s a config option that allows you to wrap the favoriteName with double quotes
```
1="Rock",2="Folk",3="Classic Alternative"
```
- only favorites at the root level of the LMS favorites hierarchy are returned
- the server handler sends the favorites list to the player handler

Add a `playFavorite` channel of type String to the squeezebox player thing.
- when you send the favoriteId to an item linked to a player's `playFavorite` channel, it plays the favorite on that player.
- the player uses the favorites list from the server handler to update the `playFavorite` channel's state description to include the favorites as options.

Dependencies: 
- Sitemap needs to be updated to be able to use the item's state options as the selection list whenever the state options change
- HABpanel needs to be able to update the choices in the Selection widget whenever the state options change

tavalin commented:
Instead of having this channel as a string of comma-separated values, wouldn't it be better for the list (received from the server) to dynamically update the state options for the channel?

mhilbush commented:
Yes, I'm actually working on that right now.  :-)  Passing the favorites list to the players, who will then update the state options on their favoritesPlay channel.  :-)  Just learning how to do that...

I still may keep the server channel the way it is, since someone might find it handy to process that in a rule.

mhilbush commented:
BTW @tavalin , would you be able to point me to a good example of how to update the StateDescription during runtime?  ;-)

tavalin commented:
The HarmonyHub binding updates the state options during runtime.  Have a look here https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java#L160

cweitkamp commented:
I added some details to eclipse/smarthome#4942. But lets continue our our discussion about the best practice how to implement that feature here.

The best way to solve this is to implement a [DynamicStateDescriptionProvider](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/DynamicStateDescriptionProvider.java) for the binding.

Have a look here for an example: https://github.com/openhab/openhab2-addons/pull/2345/files#diff-005ee22fbafd96dfe88650528963939f

mhilbush commented:
@cweitkamp That's a great example, and it seems to work very well. Much simpler than everything else I looked at. Thanks!

It doesn't look like HABpanel updates the options dynamically during runtime. I posed a question about that on the habpanel repo.

In fact, it looks a refresh of the Control page is needed in PaperUI to refresh the options.  Is that you're understanding of how it works, or how it's supposed to work?

mhilbush commented:
@cweitkamp  Your review of this would be appreciated. I followed your recommendation to use DynamicStateDescriptionProvider for managing the update to the channel's state options. Thank you for that!

cweitkamp commented:
You are welcome. I'm going to review your PR in the next few days. Thanks for asking.

> it looks a refresh of the Control page is needed in PaperUI to refresh the options. Is that you're understanding of how it works, or how it's supposed to work?

To be honest, I can't tell you. I suppose such an event doesn't exist. Paper UI has no chance to listen for an dynamically updated options list. Could be true for other UIs as well.

cweitkamp commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +867,135 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+            for (Favorite favorite : favorites) {
+                // If not quoting, we don't want any embedded commas
+                adjustedName = includeQuotes ? favorite.name : favorite.name.replaceAll(",", "");
+                sb.append(favorite.shortId + "=");
```
Avoid concatenation of strings, if you use a `StringBuilder`. Use the `append` method all the time.

cweitkamp commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +867,135 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+
+            // If the channel doesn't exist we can't update it
+            if (getThing().getChannel(CHANNEL_FAVORITES_LIST) == null) {
+                logger.debug("Channel '{}' does not exist.  Delete and readd player thing to pick up channel.",
```
Do I see double whitespaces between the sentences?

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/internal/SqueezeBoxStateDescriptionOptionsProvider.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Dynamic provider of state options while leaving other state description fields as original.
+ *
+ * @author Gregory Moyer - Initial contribution
```
I can do that. I wasn't sure the best way to handle this. Is there another way that's more preferred?

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +867,135 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+
+        private void updateChannelFavoritesList(List<Favorite> favorites) {
+            // Get config parameter indicating whether name should be wrapped with double quotes
+            final boolean includeQuotes = getConfigAs(SqueezeBoxServerConfig.class).quoteFavoritesList;
```
Was looking to provide some flexibility, depending on where/how someone was going to use this list (presumably in a rule).
```
1=Rock,2=Folk,3=Classic Alternative
```
Or
```
1="Rock",2="Folk",3="Classic Alternative"
```

cweitkamp replied:
Alright, just a guess: Is the `quoteFavoritesList` configuration more a channel configuration than a thing configuration? Is it? But I don't want you to blow up this PR with it.

mhilbush replied:
Well, yeah, I actually thought about that after I implemented this way.  It probably should be a channel configuration, huh?

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxPlayerHandler.java:
```diff
@@ -39,10 +41,13 @@
+import org.eclipse.smarthome.core.types.StateOption;
 import org.eclipse.smarthome.core.types.UnDefType;
 import org.eclipse.smarthome.io.net.http.HttpUtil;
 import org.openhab.binding.squeezebox.SqueezeBoxBindingConstants;
```
Yes. It's used in `updateFavoritesListEvent` to create the `arraylist` for `setStateOptions`.

cweitkamp replied:
You removed the reference in one of your last commits. And as you already imported it `static` it's redundant.

mhilbush replied:
Yeah, there was another reference that I missed.  I'll fix it.

mhilbush commented:
@cweitkamp  Thanks for the feedback; very much appreciated.

I believe I addressed all your feedback except for where I added comments/questions above.

That double-space between sentences thing...  Showing my age. :-(

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/internal/SqueezeBoxStateDescriptionOptionsProvider.java:
```diff
@@ -0,0 +1,61 @@
+    public void deactivate() {
+        channelOptionsMap.clear();
+    }
+}
```
I have no idea why the newline is not showing up. I added it, and I can see it clearly in the IDE.

cweitkamp commented:
You're welcome.

> That double-space between sentences thing... Showing my age. :-(

Don't bother. You are doing a good job and I really enjoy working together with you. I thought twice about  commenting that line or not. ;-)

mhilbush commented:
@martinvw This is ready for final review. Thanks.

mhilbush commented:
@cweitkamp So what do you think about moving the config to the channel? If that's the more correct approach, then it would be better to do now. WDYT?

cweitkamp commented:
I'm pretty sure you can foresee my answer...

mhilbush commented:
Yeah, it was somewhat of a rhetorical question. LOL

It's already done, so I just need to push it.  :-)

cweitkamp commented:
That was quick. Some minor remarks.

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/ESH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +98,27 @@
+		<description>Comma-separated list of favorites of form favoriteId=favoriteName</description>
+		<state readOnly="true" pattern="%s"></state>
+		<config-description>
+			<parameter name="quoteFavoritesList" type="boolean" required="true">
```
Ok

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/README.md:
````diff
@@ -48,39 +48,56 @@ Bridge squeezebox:squeezeboxserver:myServer [ ipAddress="192.168.1.10", webport=
 ```
 
-## Channels
+## Server Channels
````
Yep. Missed that.

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +869,134 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+        }
+
+        private void updateChannelFavoritesList(List<Favorite> favorites) {
+            final Channel channel = thing.getChannel(CHANNEL_FAVORITES_LIST);
```
Ok. But why does BaseThingHandler expose both?

cweitkamp replied:
I have to admit. This was nitpicking. Sry for that.

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/internal/config/SqueezeBoxServerConfig.java:
```diff
@@ -40,4 +40,8 @@
+    /*
+     * When true, wrap the name of the favorite in quotes
+     */
+    public boolean quoteFavoritesList;
```
Ok

mhilbush commented:
Thanks for the help @cweitkamp!

@martinvw Ready for final review

martinvw commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -370,6 +390,15 @@ private void disconnect() {
+        logger.trace("Updating channel {} for {} to state {}", channelID, getThing().getUID(), state);
+        try {
+            updateState(channelID, state);
+        } catch (IllegalStateException e) {
```
In what case will this happen of did I already ask that?

mhilbush replied:
The API spec for `BaseThingHandler` says `if handler is not initialized correctly, because no callback is present`

And, I chose to not do this. ;-)
https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxPlayerHandler.java#L471

martinvw replied:
Hmm, I'm not happy with this, @sjka  @triller-telekom do you have any useful suggestions here.

triller-telekom replied:
These [lines](https://github.com/eclipse/smarthome/blob/7717a6de4b4bd68eb0b6d02f7312ed2eb6ee57f0/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L343-L347) are in the `BaseThingHandler` since 17.03.2015. So this Exception was thrown ever since then. Since it is a `RuntimeException` it does not have to be caught at compile time. And since all the other bindings do not catch it at compile time I think it's alright to let it pass through to the code above.

mhilbush replied:
Ok. I'll remove it from here and `SqueezeBoxPlayerHandler`.

martinvw replied:
Did you do that already?

mhilbush replied:
Duh. It would help if I actually pushed the commit.

martinvw commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +869,134 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+                includeQuotes = (Boolean) channel.getConfiguration().get(CHANNEL_CONFIG_QUOTE_LIST);
+            }
+
+            final String quote = includeQuotes.booleanValue() ? "\"" : "";
```
I prefer to leave away the `final` here since they are not consistently used (in this binding) anyways.

mhilbush replied:
Ok

martinvw commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/internal/SqueezeBoxStateDescriptionOptionsProvider.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public @Nullable StateDescription getStateDescription(@NonNull Channel channel, @Nullable StateDescription original,
```
CC @triller-telekom if I remember correctly we should not be using both of these in this case it should 

See also https://www.eclipse.org/smarthome/documentation/development/guidelines.html#a-code-style A.12

mhilbush replied:
@triller-telekom Can you also weigh in on this?

cweitkamp replied:
I can jump in here: In general we want to avoid using @NonNull but use @NonNullByDefault on class level. You can have a look at #3207 I already changed adopted it in my PR.

triller-telekom replied:
@mhilbush Yes I second @cweitkamp's comment, we are only using `@NonNullByDefault` and `@Nullable`, see the link that @martinvw has posted.

martinvw commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +869,134 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+            }
+
+            List<Favorite> favorites = new ArrayList<>();
+            String title = "";
```
Where is this used for?

mhilbush replied:
I chose to capture the favorites title and count, as they are returned in the response from the LMS, even though they currently are not used anywhere. Would you rather they be removed, or commented as currently unused?

martinvw replied:
Or you can log them inside a `isTraceEnabled` and not store it at all?

martinvw commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +869,134 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+
+            List<Favorite> favorites = new ArrayList<>();
+            String title = "";
+            String count = "";
```
Where is this used for?

mhilbush replied:
See my comment above.

martinvw commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +869,134 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+                // Favorite type
+                else if (part.startsWith("type%3A")) {
+                    type = decode(part.substring("type%3A".length()));
+                    if (f != null) {
```
Is the order guaranteed, will we revisit this if `f` is `null`?

mhilbush replied:
Guaranteed is a little strong. :-) 

The order is defined in the LMS code, and is highly unlikely to change at this point. The convention in the LMS code generally is that the id is the first value returned in a response. I was simply trying to be defensive in case of a corrupted response.

The response from the LMS looks like this.
```
favorites = title count favorite1 favorite2 ...

where favoriteN = id name type isaudio hsitems
```

martinvw commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +869,134 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+            String count = "";
+            Favorite f = null;
+            for (String part : messageParts) {
+                String id;
```
Why do you need all these fields a quick looks suggest that you either assign them immediately or discard the data, wdyt?

mhilbush replied:
Same as my comment above. They are returned in the response from the LMS, even though they currently are not used.

martinvw commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -839,14 +869,100 @@ public void updateListener(SqueezeBoxPlayerEventListener listener) {
+
+            List<Favorite> favorites = new ArrayList<>();
+            Favorite f = null;
+            for (String part : messageParts) {
```
I did actually not mean to remove all the fields but the big benefit is that this method now more or less fits on my screen :-)

mhilbush replied:
Yes, I know. But this is actually more consistent with how the binding handles other LMS messages. ;-)

mhilbush commented:
@martinvw I think I've addressed your comments.

I do have a couple more questions.
- Is there a more preferred way [here](https://github.com/openhab/openhab2-addons/pull/3190/files#diff-6abbe31c933b849f493ec1a48d6a5e0eR958) to handle the case where the favorites list is empty?
- Perhaps for another PR, but what is the guidance for which classes should be in the `org.openhab.binding.squeezebox.handler` package?

martinvw commented:
Thanks all!

mhilbush commented:
Thank you @martinvw and @cweitkamp.

staehler commented:
@mhilbush Could you please provide an example how to use favoritList and playFavorit. I would like to use this feature :-)
Thank you very much.

mhilbush commented:
>Could you please provide an example

@staehler  I added some directions here.
https://community.openhab.org/t/squeezebox-favorites/39355/12

triller-telekom commented:
@mhilbush How about a follow-up PR where you add the instructions to the official README.md of the binding? :)

mhilbush commented:
@triller-telekom Good idea. I'm working on a PR to replace the notification volume channel with a config parameter. I'll also add the favorites instructions to the README.
